### ISSUE TITLE
[Render] Particle rendered as full-bright for a few frames on item/experience pickup

### ISSUE THREAD

**Issue by yezhiyi9670** [MinecraftForge/MinecraftForge#6706]
**Minecraft Version:** 1.15.2

**Forge Version:** 1.15.2-31.2.0

**Logs:** [latest.log with a forced crash](https://github.com/MinecraftForge/MinecraftForge/files/4650997/latest.log)

**Steps to Reproduce:**
 1. Create a world. Set to night, rainy weather and survival mode.
 2. Dig some dirt. **Note that either the rain splash particles or block destruction particles (or both) are rendered as full-bright for a few frames**.
 3. If this doesn't work, try instantly restarting the game.

**Additional Information:**

![image](https://user-images.githubusercontent.com/39194383/82336836-9f908880-9a1d-11ea-8085-7ee8d06859e3.png)

~~Maybe Forge rendering has exploited [MC-163575](https://bugs.mojang.com/browse/MC-163575). In this case, fixing MC-163575 will close this issue.~~

This can also happen after Optifine fixed MC-163575, but the issue can be still related to that.

**Comment by yezhiyi9670** [comment]
Can anyone reproduce it? It's quite long-standing without being reported before.

**Comment by yezhiyi9670** [comment]
~~A fix for [MC-168672](https://bugs.mojang.com/browse/MC-168672) is found in `net.minecraft.client.particle.ParticleManager#renderParticles`, prehaps the fix is broken and causes this issue. I'll check it later.~~

I have fixed the issue.

This issue is related to MC-168672, however, Forge didn't fix it completely (this is likely a stupid mistake).

```java
@Mixin(ParticleManager.class)
abstract public class MixinParticleManager {
    @Shadow @Final
    private Map<IParticleRenderType, Queue<Particle>> byType = Maps.newIdentityHashMap();

    @Shadow @Final
    private TextureManager renderer;

    @Overwrite
    public void renderParticles(MatrixStack matrixStackIn, IRenderTypeBuffer.Impl bufferIn, LightTexture lightTextureIn, ActiveRenderInfo activeRenderInfoIn, float partialTicks) {
        lightTextureIn.enableLightmap();
        Runnable enable = () -> {
            RenderSystem.enableAlphaTest();
            RenderSystem.defaultAlphaFunc();
            RenderSystem.enableDepthTest();
            RenderSystem.enableFog();
        };
        RenderSystem.pushMatrix();
        RenderSystem.multMatrix(matrixStackIn.getLast().getMatrix());

        for(IParticleRenderType iparticlerendertype : this.byType.keySet()) { // Forge: allow custom IParticleRenderType's
            if (iparticlerendertype == IParticleRenderType.NO_RENDER) continue;
            enable.run(); //Forge: MC-168672 Make sure all render types have the correct GL state.
            lightTextureIn.enableLightmap(); //Forge is right, but made a mistake.
            Iterable<Particle> iterable = this.byType.get(iparticlerendertype);
            if (iterable != null) {
                RenderSystem.color4f(1.0F, 1.0F, 1.0F, 1.0F);
                Tessellator tessellator = Tessellator.getInstance();
                BufferBuilder bufferbuilder = tessellator.getBuffer();
                iparticlerendertype.beginRender(bufferbuilder, this.renderer);

                for(Particle particle : iterable) {
                    try {
                        particle.renderParticle(bufferbuilder, activeRenderInfoIn, partialTicks);
                    } catch (Throwable throwable) {
                        CrashReport crashreport = CrashReport.makeCrashReport(throwable, "Rendering Particle");
                        CrashReportCategory crashreportcategory = crashreport.makeCategory("Particle being rendered");
                        crashreportcategory.addDetail("Particle", particle::toString);
                        crashreportcategory.addDetail("Particle Type", iparticlerendertype::toString);
                        throw new ReportedException(crashreport);
                    }
                }

                iparticlerendertype.finishRender(tessellator);
            }
        }

        RenderSystem.popMatrix();
        RenderSystem.depthMask(true);
        RenderSystem.disableBlend();
        RenderSystem.defaultAlphaFunc();
        lightTextureIn.disableLightmap();
        RenderSystem.disableFog();
    }
}

```

Forge made sure that all render types have the correct GL state, however, forgot to make sure that all render type have the lightmap enabled. There is a chance when item pickup particles are rendered before other types, breaking the rendering of following particles.

**Comment by yezhiyi9670** [comment]
Actually this does not occur in Vanilla, because in Vanilla the rendering order of different particle types is constant.

```java
for(IParticleRenderType iparticlerendertype : this.byType.keySet()) { // Forge: allow custom IParticleRenderType's
```

**Comment by gigaherz** [comment]
Should be fixed as of fc3fa13bf3d.
